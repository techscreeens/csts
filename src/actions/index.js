export const ADD_TILE = "ADD_TILE";
export const DELETE_TILE = "DELETE_TILE";

export const UPDATE_FIELD_VALUE = "UPDATE_FIELD_VALUE";

export const UPDATE_ORDER_BY = "UPDATE_ORDER_BY";

export const addTile = () => ({
  type: ADD_TILE
});

export const deleteTile = id => ({
  type: DELETE_TILE,
  payload: {
    id
  }
});

export const updateFieldValue = ({ id, name, value }) => ({
  type: UPDATE_FIELD_VALUE,
  payload: {
    id,
    name,
    value
  }
});

export const updateOrderBy = orderBy => ({
  type: UPDATE_ORDER_BY,
  payload: {
    orderBy
  }
});
