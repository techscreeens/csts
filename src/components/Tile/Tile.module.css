@import '../../utils/Colors.css';

.root {
  display: flex;
  flex-direction: column;

  border: 1px solid var(--border);
  box-shadow: var(--blue-shadow);

  max-width: 300px;
}

.root .content {
  flex: 1 1 auto;
  display: flex;
  flex-direction: column;

  padding: 20px;
}

.root .content textarea {
  resize: none;
}

.root .content > div:not(:last-of-type) {
  margin-bottom: 20px;
}

.root .footer {
  flex: 0 0 50px;
  display: flex;
  justify-content: space-between;
  align-items: center;

  font-size: 0.857rem;
  line-height: 1;

  padding-left: 20px;
}

.root .footer .info {
  flex: 0 0 auto;

  color: var(--subtitle);
  font-size: 0.857rem;
}

.root .footer .info p {
  margin: 3px 0;
}

.root .footer .confirmDelete {
  color: var(--destructive);
  user-select: none;
}

.root .footer .delete {
  display: flex;
  justify-content: center;
  align-items: center;

  width: 40px;
  height: 40px;

  color: var(--destructive);
  cursor: pointer;
}

.root .footer .delete:hover {
  color: var(--destructive-hover);
}
