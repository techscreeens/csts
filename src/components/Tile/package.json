{
  "name": "Tile",
  "version": "1.0.0",
  "main": "./Tile.js"
}
