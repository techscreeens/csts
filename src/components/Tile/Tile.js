import React, { memo, useCallback, useEffect, useState } from "react";
import cx from "classnames";

import { deleteTile, updateFieldValue } from "../../actions";
import { DATETIME_SHORT_WITH_SECONDS } from "../../utils/DateTimeFormatters";

import { TextArea, TextInput } from "../Form";
import { Delete } from "../Icons";

import s from "./Tile.module.css";

export default memo(function Tile({
  tile: { id, title, description, createdAt, updatedAt },
  className,
  dispatch
}) {
  const [isDeleting, setIsDeleting] = useState(false);

  useEffect(() => {
    if (isDeleting) {
      const timeout = setTimeout(() => setIsDeleting(false), 3000);
      return () => clearTimeout(timeout);
    }
  }, [isDeleting, setIsDeleting]);

  const onFieldChange = useCallback(
    ({ currentTarget: { name, value } }) =>
      dispatch(updateFieldValue({ id, name, value })),
    [id, dispatch]
  );

  const onDelete = useCallback(() => {
    if (isDeleting) {
      dispatch(deleteTile(id));
    } else {
      setIsDeleting(true);
    }
  }, [id, isDeleting, dispatch]);

  return (
    <div className={cx(s.root, className)}>
      <div className={s.content}>
        <TextInput
          name="title"
          label="Title"
          value={title}
          onChange={onFieldChange}
          autoFocus={!title}
        />
        <TextArea
          name="description"
          label="Description"
          description={description}
          value={description}
          maxLength={140}
          onChange={onFieldChange}
          autoExpand
        />
      </div>
      <div className={s.footer}>
        {!isDeleting && (
          <div className={s.info}>
            <p>
              <span>Created: </span>
              {createdAt.toLocaleString([], DATETIME_SHORT_WITH_SECONDS)}
            </p>
            <p>
              <span>Updated: </span>
              {updatedAt.toLocaleString([], DATETIME_SHORT_WITH_SECONDS)}
            </p>
          </div>
        )}
        {isDeleting && (
          <div className={s.confirmDelete}>Click again to confirm</div>
        )}
        <div
          className={s.delete}
          onClick={onDelete}
          role="button"
          aria-label="Delete"
        >
          <Delete />
        </div>
      </div>
    </div>
  );
});
