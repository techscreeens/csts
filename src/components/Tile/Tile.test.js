import React from "react";
import { render, fireEvent, act } from "react-testing-library";

import { updateFieldValue, deleteTile } from "../../actions";

import Tile from "./Tile";

describe("Tile", () => {
  const dispatchMock = jest.fn();
  const tile = {
    id: "8ea381a4-a477-4258-b2fd-8a23e544457e",
    title: "Title Field",
    description: "Description Field",
    createdAt: new Date("2019-05-05T12:00:00.000Z"),
    updatedAt: new Date("2019-05-06T12:00:00.000Z")
  };

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should render correctly", () => {
    const { asFragment } = render(
      <Tile className="testClassName" tile={tile} dispatch={dispatchMock} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have a title field", () => {
    const { getByLabelText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );
    expect(() => getByLabelText("Title")).not.toThrow();
  });

  it("should have a description field", () => {
    const { getByLabelText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );
    expect(() => getByLabelText("Description")).not.toThrow();
  });

  it("should have a created date field", () => {
    const { getByText } = render(<Tile tile={tile} dispatch={dispatchMock} />);
    expect(() => getByText("Created:")).not.toThrow();
  });

  it("should have an updated date field", () => {
    const { getByText } = render(<Tile tile={tile} dispatch={dispatchMock} />);
    expect(() => getByText("Updated:")).not.toThrow();
  });

  it("should have a max limit of 140 characters in the description field", () => {
    const { getByLabelText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );
    const description = getByLabelText("Description");
    expect(description).toHaveAttribute("maxLength", "140");
  });

  it("should dispatch an updateFieldValue action when the title field is typed into", () => {
    const { getByLabelText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );

    const title = getByLabelText("Title");
    fireEvent.change(title, { target: { name: "title", value: "abc" } });

    expect(dispatchMock).toHaveBeenCalledTimes(1);
    expect(dispatchMock).toHaveBeenCalledWith(
      updateFieldValue({
        id: tile.id,
        name: title.name,
        value: "abc"
      })
    );
  });

  it("should dispatch an updateFieldValue action when the description field is typed into", () => {
    const { getByLabelText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );

    const description = getByLabelText("Description");
    fireEvent.change(description, {
      target: { name: "Description", value: "def" }
    });

    expect(dispatchMock).toHaveBeenCalledTimes(1);
    expect(dispatchMock).toHaveBeenCalledWith(
      updateFieldValue({
        id: tile.id,
        name: description.name,
        value: "def"
      })
    );
  });

  it("should show confirmation text when the delete button is pressed for the first time", () => {
    const { getByLabelText, getByText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );
    fireEvent.click(getByLabelText("Delete"));

    expect(() => getByText("Click again to confirm")).not.toThrow();
    expect(dispatchMock).not.toHaveBeenCalled();
  });

  it("should delete the idea when the delete button is pressed while the confirmation text is displayed", () => {
    const { getByLabelText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );
    fireEvent.click(getByLabelText("Delete"));
    fireEvent.click(getByLabelText("Delete"));

    expect(dispatchMock).toHaveBeenCalledTimes(1);
    expect(dispatchMock).toHaveBeenCalledWith(deleteTile(tile.id));
  });

  it("should reset the deletion state after 3 seconds", () => {
    jest.useFakeTimers();

    const { getByLabelText, getByText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );
    fireEvent.click(getByLabelText("Delete"));

    expect(() => getByText("Click again to confirm")).not.toThrow();

    act(() => {
      jest.advanceTimersByTime(3000);
      jest.runAllTimers();
    });
    expect(() => getByText("Click again to confirm")).toThrow();
  });

  it("should focus the title field on mount if title is empty", () => {
    const { getByLabelText } = render(
      <Tile tile={{ ...tile, title: "" }} dispatch={dispatchMock} />
    );

    expect(getByLabelText("Title")).toHaveFocus();
  });

  it("should not focus the title field on mount if title isn't empty", () => {
    const { getByLabelText } = render(
      <Tile tile={tile} dispatch={dispatchMock} />
    );

    expect(getByLabelText("Title")).not.toHaveFocus();
  });
});
