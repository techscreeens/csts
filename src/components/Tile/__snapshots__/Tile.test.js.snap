// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile should render correctly 1`] = `
<DocumentFragment>
  <div
    class="root testClassName"
  >
    <div
      class="content"
    >
      <div
        class="root"
      >
        <label
          for="title"
        >
          Title
        </label>
        <div
          class="input input"
        >
          <input
            id="title"
            name="title"
            value="Title Field"
          />
        </div>
      </div>
      <div
        class="root"
      >
        <label
          for="description"
        >
          Description
        </label>
        <div
          class="input input"
        >
          <textarea
            id="description"
            maxlength="140"
            name="description"
            style="height: 0px;"
          >
            Description Field
          </textarea>
          <div
            class="measurer"
          >
            Description Field
          </div>
        </div>
        <div
          class="countdown"
        >
          123 characters remaining
        </div>
      </div>
    </div>
    <div
      class="footer"
    >
      <div
        class="info"
      >
        <p>
          <span>
            Created: 
          </span>
          5/5/2019, 12:00:00 PM
        </p>
        <p>
          <span>
            Updated: 
          </span>
          5/6/2019, 12:00:00 PM
        </p>
      </div>
      <div
        aria-label="Delete"
        class="delete"
        role="button"
      >
        <div
          class="root"
        >
          <svg
            fill="currentColor"
            height="20"
            viewBox="0 0 32 32"
            width="20"
          >
            <path
              d="M 15 4 C 14.476563 4 13.941406 4.183594 13.5625 4.5625 C 13.183594 4.941406 13 5.476563 13 6 L 13 7 L 7 7 L 7 9 L 8 9 L 8 25 C 8 26.644531 9.355469 28 11 28 L 23 28 C 24.644531 28 26 26.644531 26 25 L 26 9 L 27 9 L 27 7 L 21 7 L 21 6 C 21 5.476563 20.816406 4.941406 20.4375 4.5625 C 20.058594 4.183594 19.523438 4 19 4 Z M 15 6 L 19 6 L 19 7 L 15 7 Z M 10 9 L 24 9 L 24 25 C 24 25.554688 23.554688 26 23 26 L 11 26 C 10.445313 26 10 25.554688 10 25 Z M 12 12 L 12 23 L 14 23 L 14 12 Z M 16 12 L 16 23 L 18 23 L 18 12 Z M 20 12 L 20 23 L 22 23 L 22 12 Z "
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
